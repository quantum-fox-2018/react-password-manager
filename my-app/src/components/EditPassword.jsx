import React, { Component } from 'react';
import {editPassword, getPassword} from '../store/password/password.action'
import {connect} from 'react-redux'
import {bindActionCreators} from 'redux'

class EditPassword extends Component {
  constructor () {
    super ()
    this.state = {
      editId: '',
      editUrl: '',
      editUsername:'',
      editPassword: ''
    }
  }

  componentDidMount () {
    this.props.getPassword()
    let listPassword = this.props.myPassword.data
    listPassword.map(value => {
      if(value.id === this.props.match.params.passId) {
        this.setState({
          editId: value.id,
          editUrl: value.url,
          editUsername: value.username,
          editPassword: value.password
        })
      }
    })
  }

  handleChange = (e) => {
    e.preventDefault()
    this.setState({
      [e.target.name] : e.target.value
    })
  }

  editPasswordButton = () => {
    let newPass = {
      id: this.state.editId,
      url : this.state.editUrl,
      username: this.state.editUsername,
      password: this.state.editPassword
    }
    let user = localStorage.getItem('username')
    this.props.editPassword(newPass, user)
    this.props.history.push('/home')
  }
  lowerCase () {
    return (this.state.editPassword.match(/^(?=.*[a-z])/))
  }
  
  upperCase () {
    return (this.state.editPassword.match(/^(?=.*[A-Z])/)) 
  }
  
  oneNum () {
    return (this.state.editPassword.match(/^(?=.*[0-9])/))
  }
  
  minLength () {
    return (this.state.editPassword.length > 6) ? true : false
  }

  specialChar () {
    return (this.state.editPassword.match(/^(?=.*[_\W])/))
  }

  allValid() {
    return (this.minLength() && this.upperCase() && this.lowerCase() && this.oneNum() && this.specialChar())
  }
  render() {
    return (
      <div className="container"> 
        <h3>Edit data password for {this.props.myPassword.data.id}</h3>
          <form>
            <div className="form-group">
              <label>URL</label>
              <input type="text" className="form-control" name="editUrl"
              value={this.state.editUrl}
              onChange={this.handleChange}
              />
            </div>
            <div className="form-group">
              <label>Username</label>
              <input type="text" className="form-control" name="editUsername"
                value={this.state.editUsername}
                onChange={this.handleChange}
              />
            </div>
            <div className="form-group">
              <label>Password</label>
              <input type="password" className="form-control" name="editPassword"
                value={this.state.editPassword}
                onChange={this.handleChange}
              />
            </div>
            <button type="button" className={"btn btn-primary "+ (this.allValid() ? "active" : "disabled")} onClick={this.editPasswordButton}>Submit</button>
          </form>
          <div style={{marginTop:'10px'}}>
          <h4>Password requirements: </h4>
            <ul className="text-left" style={{listStyle: "none"}}>
              <li className={"alert alert-dismissible " + (this.lowerCase() ? "alert-success" : "alert-danger")}> {this.lowerCase() ? '[ O ]': '[ X ]'} At least <strong>one lowercase</strong></li>
              <li className={"alert alert-dismissible " + (this.upperCase() ? "alert-success" : "alert-danger")}> {this.upperCase() ? '[ O ]': '[ X ]'} At least <strong>one capital</strong></li>
              <li className={"alert alert-dismissible " + (this.oneNum() ? "alert-success" : "alert-danger")}> {this.oneNum() ? '[ O]': '[ X ]'} At least <strong>one number</strong></li>
              <li className={"alert alert-dismissible " + (this.minLength() ? "alert-success" : "alert-danger")}> {this.minLength() ? '[ O ]': '[ X ]'} Minimum <strong>6 characters</strong></li>
              <li className={"alert alert-dismissible " + (this.specialChar() ? "alert-success" : "alert-danger")}> {this.specialChar() ? '[ O ]': '[ X ]'} At least <strong>Use one symbol</strong></li>
            </ul>
          </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  myPassword: state.passManager
})

const mapDispatchToProps = (dispatch) => bindActionCreators({
  getPassword,
  editPassword,
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps) (EditPassword);