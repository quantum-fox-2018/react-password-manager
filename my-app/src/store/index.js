import {createStore, applyMiddleware, combineReducers} from 'redux'
import thunk from 'redux-thunk'
import passwordReducers from './password/password.reducers'
import userReducers from './users/user.reducers'
import logger from 'redux-logger'


const reducers = combineReducers({
  passManager : passwordReducers,
  user: userReducers
})

const store = createStore(
  reducers,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
  applyMiddleware(thunk, logger)
)

export default store