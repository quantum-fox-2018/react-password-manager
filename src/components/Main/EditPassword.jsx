import React, { Component } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { EditPassword as EditPassAction } from '../../store/password/password.action'
import PasswordValidation from './PasswordValidation'

class EditPassword extends Component {
	constructor() {
		super()
		this.state = {
			editId: '',
			editUrl: '',
			editEmail: '',
			editPass: '',
			upperCaseE: false,
      lowerCaseE: false,
      specialCharE: false,
      numberE: false,
      minLengthE: false,
      isItValidE: false
		}
	}

	componentDidMount() {
		this.setState({
			editId: this.props.data.id,
			editUrl: this.props.data.url,
			editEmail: this.props.data.email,
			editPass: this.props.data.password
		}, () => {
			this.uCaseValidation()
      this.lCaseValidation()
      this.sCharValidation()
      this.numberValidation()
      this.lengthValidation()
		})
	}

	handleChange = (e) => {
		this.setState({
			[e.target.name]: e.target.value
		})
	}

	pwdHandleChange = (e) => {
		this.setState({
			editPass: e.target.value
		}, () => {
			this.uCaseValidation()
      this.lCaseValidation()
      this.sCharValidation()
      this.numberValidation()
      this.lengthValidation()
		})
	}

	uCaseValidation = () => {
    let upperCase = /[A-Z]/g.test(this.state.editPass)
    if (upperCase) {
      this.setState({ upperCaseE: true }, () => this.checkAllValidation())
    } else {
      this.setState({ upperCaseE: false })
    }
    // console.log('upperCase ', upperCase)
  }

  lCaseValidation = () => {
    let lowerCase = /[a-z]/g.test(this.state.editPass)
    if (lowerCase) {
      this.setState({ lowerCaseE: true }, () => this.checkAllValidation())
    } else {
      this.setState({ lowerCaseE: false })
    }
    // console.log('lowerCase ', lowerCase)
  }

  sCharValidation = () => {
    // \W => nonword char
    let specialChar = /\W/g.test(this.state.editPass)
    if (specialChar) {
      this.setState({ specialCharE: true }, () => this.checkAllValidation())
    } else {
      this.setState({ specialCharE: false })
    }
    // console.log('special char ', specialChar)
  }

  numberValidation = () => {
    // \d digit char
    let number =  /\d/g.test(this.state.editPass)
    if (number) {
      this.setState({ numberE: true }, () => this.checkAllValidation())
    } else {
      this.setState({ numberE: false })
    }
    // console.log('number ', number)
  }

  lengthValidation = () => {
    if (this.state.editPass.length < 6) {
      // console.log('length', false)
      this.setState({ minLengthE: false })
    } else {
      // console.log('length', true)
      this.setState({ minLengthE: true }, () => this.checkAllValidation())
    }
  }

  checkAllValidation = () => {
    let editButton = document.querySelector('#submit'+this.props.data.id)
    let upperCase = this.state.upperCaseE
    let lowerCase = this.state.lowerCaseE
    let specialChar = this.state.specialCharE
    let number = this.state.numberE
    let minLength = this.state.minLengthE
    let url = this.state.editUrl
    let email = this.state.editEmail 

    if (upperCase && lowerCase && specialChar &&
    number && minLength) {
      if (editButton&& url && email) {
        editButton.classList.remove('disabled')
      }
      this.setState({isItValidE: true})
    } else {
      if (editButton&& url && email) {
        editButton.classList.add('disabled')
      }
      this.setState({ isItValidE: false})
    }
  }

  resetState = () => {
    this.setState({
      ...this.state,
      editUrl: '',
      editEmail: '',
      editPass: ''
    })
    // let url = document.getElementById('editUrl')
    // let email = document.getElementById('editEmail')
    // let pass = document.getElementById('editPass')
    // if ( url && email && pass ) {
    //   url.value = ''
    //   email.value = ''
    //   pass.value = ''
    // }
  }

	submitChange = (e) => {
		e.preventDefault()
		let userId = this.props.user.userId
		let changes = {
			...this.props.data,
			url: this.state.editUrl,
			email: this.state.editEmail,
			password: this.state.editPass
		}
		this.props.EditPassAction(changes, userId)
	}

	render() {
		const { id } = this.props.data
		let PassVal = ''
		if (!this.state.isItValidE) {
			PassVal = (
			<PasswordValidation 
				upperCase={ this.state.upperCaseE }
				lowerCase={ this.state.lowerCaseE }
				specialChar={ this.state.specialCharE }
				number={ this.state.numberE }
				minLength={ this.state.minLengthE }
			/>)
		}

		return (
			<div id={"edit"+id} className="modal fade" role="dialog">
				<div className="modal-dialog">
					<div className="row">
						<form className="col s12" onSubmit={this.submitChange}>
							<h4 className="header2">Edit Password</h4>
							<div className="row">
								<div className="input-field col s12">
									<input id={"editUrl"+id} type="text" name="editUrl"
									className="validate" value={this.state.editUrl}
                  onChange={this.handleChange} required/>
									<label htmlFor={"editUrl"+id}>Url</label>
								</div>
								<div className="input-field col s12">
									<input id={"editEmail"+id} type="text" name="editEmail"
									className="validate" value={this.state.editEmail}
                  onChange={this.handleChange} required/>
									<label htmlFor={"editEmail"+id}>Email / Username</label>
								</div>
								<div className="input-field col s12">
									<input id={"editPass"+id} type="password" name="editPass"
									className="validate" value={this.state.editPass}
                  onChange={this.pwdHandleChange} required/>
									<label htmlFor={"editPass"+id}>Password</label>
								</div>
								<div className="input-field col s12">
									{PassVal}
								</div>
								<div className="input-field col s12">
									<button id={"submit"+id} type="submit" className="btn teal waves-effect waves-light">Edit</button>
									<button className="btn teal" data-dismiss="modal">Cancel</button>
								</div>
							</div>
						</form>
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = (state) => ({
  user: state.user,
  passwords: state.passwords
})

const mapDispatchToProps = (dispatch) => bindActionCreators({
  EditPassAction
}, dispatch)


export default connect(
	mapStateToProps,
	mapDispatchToProps
)(EditPassword)