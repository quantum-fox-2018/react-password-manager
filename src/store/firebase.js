import firebase from 'firebase'

const config = {
  apiKey: "AIzaSyCJ0Hlgag0DOOKpxzUsGTaE4FAu0tIE_EE",
  authDomain: "react-password-manager-bef30.firebaseapp.com",
  databaseURL: "https://react-password-manager-bef30.firebaseio.com",
  projectId: "react-password-manager-bef30",
  storageBucket: "react-password-manager-bef30.appspot.com",
  messagingSenderId: "827996324966"
}
firebase.initializeApp(config)
export const Database = firebase.database()
export const User = firebase.auth()