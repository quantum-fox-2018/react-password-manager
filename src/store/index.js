import { createStore, applyMiddleware, combineReducers } from "redux"
import logger from 'redux-logger'
import usersReducers from './users/reducers'
import thunk from 'redux-thunk'
import passmanagerReducer from "./passmanager/reducers";

const reducers = combineReducers({
  users: usersReducers,
  passmanager: passmanagerReducer
})

let store = createStore(reducers, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(), applyMiddleware(thunk, logger))

export default store