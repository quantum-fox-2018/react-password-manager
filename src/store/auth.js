import { observable } from 'mobx'
import swal from 'sweetalert2'
import { auth } from './firebase'


class Auth {
  userId = ''
  email = ''
  @observable statusAuth = false

  errorNotif(err) {
    swal({
      position: 'center',
      type: 'error',
      title: 'Check your password',
      text: err.message
    })
  }

  successNotif(info) {
    swal({
      position: 'center',
      type: 'success',
      title: info,
      showConfirmButton: false,
      timer: 1000
    })
  }

  signUp(email, password) {
    auth.createUserWithEmailAndPassword(email, password)
    .then(res => {
      this.email = email
      this.userId = res.uid
      this.statusAuth = true
      this.successNotif('Enjoy!')
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }

  signIn(email, password) {
    auth.signInWithEmailAndPassword(email, password)
    .then(res => {
      this.email = email
      this.userId = res.uid
      this.statusAuth = true
      this.successNotif('Welcome Back!')
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }

  signOut() {
    this.email = ''
    this.userId = ''
    auth.signOut()
    .then(res => {
      this.successNotif('See You!')
      this.statusAuth = false
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }
}

export default new Auth()
