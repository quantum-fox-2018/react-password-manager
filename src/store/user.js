import { observable } from 'mobx'
import swal from 'sweetalert2'


import { db, auth } from './firebase'
import authData from './auth'

class PassManager {
  @observable listData = []
  valUrl = false
  valLength = false
  valNumber = false
  valUpper = false
  valLower = false
  valSpecial = false
  allVal = false

  checkUrl(url) {
    let regex = /[-a-zA-Z0-9@:%_+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_+.~#?&//=]*)?/gi
    this.valUrl = regex.test(url)
    return this.valUrl
  }

  checkLength(password) {
    this.valLength = (password.length >= 6)
    return this.valLength
  }

  checkNumber(password) {
    this.valNumber = (/\d/g.test(password))
    return this.valNumber
  }

  checkUpper(password) {
    this.valUpper = (/[A-Z]/g.test(password))
    return this.valUpper
  }

  checkLower(password) {
    this.valLower = (/[a-z]/g.test(password))
    return this.valLower
  }

  checkSpecial(password) {
    this.valSpecial = (/\W/g.test(password))
    return this.valSpecial
  }

  checkAllValidation() {
    if(this.valUrl && this.valLength && this.valNumber && this.valUpper && this.valLower && this.valSpecial) {
      this.allVal = true
      return this.allVal
    }
  }

  hashPassword(str) {
    let length = str.length
    let bullet = String.fromCharCode(0x2022)
    return bullet.repeat(length)
  }

  showPassword(userId, idData) {
    swal.mixin({
      input: 'password',
      confirmButtonText: 'Confirm',
      showCancelButton: true,
    }).queue([
      {
        title: 'Password',
        text: 'enter your password'
      }
    ]).then((result) => {
      if (!result.value) {
        swal({
          type: 'question',
        })
      } else if (result.value[0]) {
        auth.signInWithEmailAndPassword(authData.email, result.value[0])
        .then(res => {
          let data = this.listData.map(data => {
            if(data.id === idData) {
              return {...data, showPassword: !data.showPassword}
            } else {
              return {...data}
            }
          })
          this.listData = data  
        })
        .catch(err => {
          swal({
            type: 'error',
            title: 'Error',
            text: err.message
          })
        })
      }
    })
  }

  loadData(userId) {
    db.ref(`/user/${userId}`).once('value', (snap) => {
      const pass = snap.val()
      let arr = []
      for (const key in pass) {
        if (pass.hasOwnProperty(key)) {
          const el = pass[key]
          const update = {...el, id: key, showPassword: false}
          arr.push(update)
        }
      }
      this.listData = arr
    })
  }

  createData(userId, url, username, password) {
    db.ref(`/user/${userId}`).push({
      url, username, password, createdAt: Date.now(), updatedAt: null
    })
    .then(res => {
      swal({
        position: 'center',
        type: 'success',
        title: 'Your password has been saved',
        showConfirmButton: false,
        timer: 1500
      })
      this.loadData(userId)
    })
    .catch(err => {
      alert(err)
    })
  }

  deleteData(userId, idData) {
    swal({
      title: 'Are you sure?',
      text: "You won't be able to revert this!",
      type: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#3085d6',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Yes, delete it!'
    }).then((result) => {
      if (result.value) {
        db.ref(`/user/${userId}`).child(idData).remove()
        this.loadData(userId)
        swal(
          'Deleted!',
          'Your file has been deleted.',
          'success'
        )
      }
    })
  }

  editPassword(userId, idData) {
    swal.mixin({
      input: 'password',
      confirmButtonText: 'Confirm',
      showCancelButton: true,
      inputAttributes: {
        'minlength': 6,
        'autocapitalize': 'off',
        'autocorrect': 'off'
      }
    }).queue([
      {
        title: 'Edit Password',
        text: 'enter your new password'
      }
    ]).then((result) => {
      if (result.value) {
        let data = this.listData.map(data => {
          if(data.id === idData) {
            let newData = {...data, password: result.value[0], updatedAt: Date.now()}
            db.ref(`/user/${userId}`).child(idData).update(newData)
            return newData
          } else {
            return {...data}
          }
        })
        this.listData = data
        swal({
          position: 'center',
          type: 'success',
          title: 'Your password has been edited',
          showConfirmButton: false,
          timer: 1500
        })
      } else {
        swal({
          type: 'question',
        })
      }
    })
  }
}

export default new PassManager()
