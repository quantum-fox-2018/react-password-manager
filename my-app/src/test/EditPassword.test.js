import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, {shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Provider} from 'react-redux'
import {Route, Switch} from 'react-router-dom'
import store from '../store'
import LocalStorageMock from './localstorageMock'
import renderer from 'react-test-renderer'
import EditPassword from '../components/EditPassword'

global.localStorage = new LocalStorageMock();
Enzyme.configure({ adapter: new Adapter() });


describe('<AddPassword/>', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    const wrapper = mount(
    <Provider store={store}>
      <EditPassword/>
    </Provider>
    )
    ReactDOM.render(wrapper, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders without crashing snapshot', () => {
    const rendered = renderer.create(
    <Provider store={store}>
      <EditPassword/>
    </Provider>)
    expect(rendered).toMatchSnapshot()
  });
})