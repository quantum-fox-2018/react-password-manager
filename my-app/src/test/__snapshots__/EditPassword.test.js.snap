// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPassword/> renders without crashing snapshot 1`] = `
<div
  className="container"
>
  <h3>
    Edit data password for 
  </h3>
  <form>
    <div
      className="form-group"
    >
      <label>
        URL
      </label>
      <input
        className="form-control"
        name="editUrl"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Username
      </label>
      <input
        className="form-control"
        name="editUsername"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Password
      </label>
      <input
        className="form-control"
        name="editPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary disabled"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </form>
  <div
    style={
      Object {
        "marginTop": "10px",
      }
    }
  >
    <h4>
      Password requirements: 
    </h4>
    <ul
      className="text-left"
      style={
        Object {
          "listStyle": "none",
        }
      }
    >
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one lowercase
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one capital
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one number
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         Minimum 
        <strong>
          6 characters
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          Use one symbol
        </strong>
      </li>
    </ul>
  </div>
</div>
`;
